### PR TITLE
Allow document attachments to be downloaded

### PR DESCRIPTION
**Pull request by eanders-MS**

### CODE REVIEW

**Comment by eanders-MS** [comment]
Related:
- https://github.com/Microsoft/BotFramework-WebChat/issues/201
- https://github.com/Microsoft/BotFramework-Emulator/issues/143

**Comment by danmarshall** [comment]
can you attach a quick UI rendering? Adding @jecristi

**Comment by eanders-MS** [comment]
From TestBot
image

**Comment by danmarshall** [comment]
Can you add the filesize below it ? Might be nice to know if you were on a mobile data plan :)

**Comment by eanders-MS** [comment]
I wish I could, but we don't have this info.

File: src/Attachment.tsx [hunk 1]
```diff
@@ -115,6 +115,25 @@ const Media = (props: {
...
+    if (regExpCard.test(props.contentType)) {
+        return <span>{ props.format.strings.unknownCard.replace('%1', props.contentType) }</span>;
+    } else if (props.contentUrl) {
+        if (props.name) {
+            return <span><a href={ props.contentUrl } title={ props.contentUrl } target='_blank'>{ props.name }</a></span>;
+        } else {
+            return <span><a href={ props.contentUrl } title={ props.contentUrl } target='_blank'>{ props.format.strings.unknownFile.replace('%1', props.contentType) }</a></span>;
+        }
+    } else {
+        return <span>{ props.format.strings.unknownFile.replace('%1', props.contentType) }</span>;
```
Comment (danmarshall) [inline src/Attachment.tsx]: any way to collapse the duplicates?

Comment (eanders-MS) [inline src/Attachment.tsx]: excellent suggestion. coming right up...

File: src/Attachment.tsx [hunk 2]
```diff
@@ -115,6 +115,21 @@ const Media = (props: {
...
+    contentType: string,
+    contentUrl: string,
+    name: string
+}) => {
+    if (regExpCard.test(props.contentType)) {
+        return <span>{ props.format.strings.unknownCard.replace('%1', props.contentType) }</span>;
+    } else if (props.contentUrl) {
+        return <span><a href={ props.contentUrl } title={ props.contentUrl } target='_blank'>{ props.name || props.format.strings.unknownFile.replace('%1', props.contentType) }</a></span>;
+    } else {
+        return <span>{ props.format.strings.unknownFile.replace('%1', props.contentType) }</span>;
```
Comment (danmarshall) [inline src/Attachment.tsx]: looks like maybe there is on more collapse?  I think if you refactored the if... else if... else ... then you can have the one fallback.

Comment (danmarshall) [inline src/Attachment.tsx]: oh, it's unknownCard vs unknownFile.

Comment (eanders-MS) [inline src/Attachment.tsx]: True, sorry, distracted by other work and not thinking it though :/

Comment (eanders-MS) [inline src/Attachment.tsx]: oh right, yep.

**Comment by billba** [comment]
Looks good to me!

**Comment by eanders-MS** [comment]
Thank you!

**Comment by upendra1588** [comment]
hi, 
I am still receiving **[File of type 'application/pdf']** instead of PDF or doc file, does anybody help me to resolve it ? I am using botframework-emulator-Setup-3.5.25 emulator.

below is the code:

[LuisIntent("Document")]
        public async Task Document(IDialogContext context, LuisResult result)
        {


            var reply = context.MakeMessage();
            Attachment attachment1 = new Attachment();
            attachment1.ContentType = "application/pdf";
            attachment1.ContentUrl = @"C:\CIOSOH-Bot\Accenture.CIO.SOH.Bot.App\Accenture.CIO.SOH.Bot.App\Documents\keyboard-shortcuts-Visual-Studio-Code.pdf";
            
            reply.Attachments.Add(attachment1);

            reply.Text = "Angular JS";

           

            await context.PostAsync(reply);
            context.Wait(MessageReceived);
        }


Thanks
Upendra

**Comment by eanders-MS** [comment]
Hi @upendra1588, The current version of the emulator references the previous release of the webchat control. We'll release an updated emulator early next week.

**Comment by upendra1588** [comment]
thanks for prompt response

**Comment by upendra1588** [comment]
Hi Eanders-MS,

I faced one more problem in getting document file  throw WebAPI in below code.

 [LuisIntent("APITest")]
        public async Task APITest(IDialogContext context, LuisResult result)
        {
            HttpClient client = new HttpClient();
            string searchValue = string.Empty;
            HttpResponseMessage test = null;
            client.BaseAddress = new Uri("http://localhost:3979/");
            client.DefaultRequestHeaders.Accept.Clear();
            client.DefaultRequestHeaders.Accept.Add(new     MediaTypeWithQualityHeaderValue("application/octet-stream"));
            HttpResponseMessage response = await client.GetAsync("api/Registration?key=doc");
            if (response.IsSuccessStatusCode)
            {
                 test = await response.Content.ReadAsHttpResponseMessageAsync();
            }

            await context.PostAsync(test);
            context.Wait(MessageReceived);
        }
 in Response i got file but getting error to post response  **await context.PostAsync(test);** can you please help me to post response ?

thanks
upendra

**Comment by upendra1588** [comment]
Hi  eanders-MS,

let me know the emulator release date, we are getting [File of type 'application/pdf'] instead of PDF or doc file.

thanks
upendra

**Comment by eanders-MS** [comment]
Hi @upendra1588, I will post here once a new emulator is available. Current plan is to make a new release early next week.

**Comment by upendra1588** [comment]
webchat

hi guys ,

webchat control is also not render pdf /doc file, can you please check it,
 [LuisIntent("Document")]
        public async Task Document(IDialogContext context, LuisResult result)
        {
            var reply = context.MakeMessage();
            Attachment attachment1 = new Attachment();
            attachment1.ContentType = "application/pdf";
            attachment1.ContentUrl = @"C:\CIOSOH-Bot\Accenture.CIO.SOH.Bot.App\Accenture.CIO.SOH.Bot.App\Documents\keyboard-shortcuts-Visual-Studio-Code.pdf";            
            reply.Attachments.Add(attachment1);
            reply.Text = "Angular JS";
            await context.PostAsync(reply);
            context.Wait(MessageReceived);
        }

**Comment by eanders-MS** [comment]
@upendra1588 The new emulator is out now. https://emulator.botframework.com
Cheers,
Eric

**Comment by upendra1588** [comment]
@eanders-MS thanks a lot :) , now it is working for both webchat and emulator.

**Comment by eanders-MS** [comment]
@upendra1588 I regret to say we're planning to release a recovery build of the Emulator that temporarily reverts this feature while we work to stabilize unrelated issues. It will be a temporary setback.

**Comment by upendra1588** [comment]
@eanders-MS , let me know when emulator and webchat channel will support to download a file which you provide in your 3.5.26 release, we need to provide this functionality in our bot framework.

**Comment by upendra1588** [comment]
Gentle Reminder !!

@eanders-MS , let me know when emulator and webchat channel will support to download a file which you provide in your 3.5.26 release, we need to provide this functionality in our bot framework.

**Pull request merged**